mne/preprocessing/ica.py

```diff
@@ -2093,3 +2110,228 @@ def _band_pass_filter(ica, sources, target, l_freq, h_freq, verbose=None):
         raise ValueError('Must specify both pass bands')
 
     return sources, target
+
+
+@verbose
+def corrmap(icas, template, threshold="auto", label=None,
+            plot=True, ch_type="eeg"):
+    """Find similar Independent Components across subjects by map similarity.
+
+    Corrmap (Viola et al. 2009 Clin Neurophysiol) identifies the best group
+    match to a supplied template. Typically, feed it a list of fitted ICAs and
+    a template IC, for example, the blink for the first subject, to identify
+    specific ICs across subjects.
+
+    The specific procedure consists of two iterations. In a first step, the
+    maps best correlating with the template are identified. In the step, the
+    analysis is repeated with the mean of the maps identified in the first
+    stage.
+
+    Outputs a list of fitted ICAs with the indices of the marked ICs in a
+    specified field.
+
+    The original Corrmap website: www.debener.de/corrmap/corrmapplugin1.html
+
+    Parameters
+    ----------
+    icas : list of mne.preprocessing.ICA
+        A list of fitted ICA objects.
+    template : tuple
+        A tuple with two elements (int, int) representing the list indices of
+        the set from which the template should be chosen, and the template.
+    threshold : "auto" | list of float | float
+        Correlation threshold for identifying ICs
+        If "auto", search for the best map by trying all correlations between
+        0.6 and 0.95. In the original proposal, lower values are considered,
+        but this is not yet implemented.
+        If list of floats, search for the best map in the specified range of
+        correlation strengths. As correlation values, must be between 0 and 1
+        If float > 0, select ICs correlating better than this.
+        If float > 1, use find_outliers to identify ICs within subjects (not in
+        original Corrmap)
+        Defaults to "auto".
+    label : None | str
+        If not None, categorised ICs are stored in a dictionary "labels_" under
+        the given name. Preexisting entries will be appended to
+        (excluding repeats), not overwritten. If None, a dry run is performed.
+    plot : bool
+        Should constructed template and selected maps be plotted?
+    ch_type : 'mag' | 'grad' | 'planar1' | 'planar2' | 'eeg'
+            The channel type to plot. Defaults to 'eeg'.
+
+    Returns
+    -------
+    template_fig : fig
+        Figure showing the mean template.
+    labelled_ics : fig
+        Figure showing the labelled ICs per subject.
+    """
+
+    def get_ica_map(ica, components=None):
+        if components is None:
+            components = list(range(ica.n_components_))
+        maps = fast_dot(ica.mixing_matrix_[:, components].T,
+                        ica.pca_components_[:ica.n_components_])
+        return maps
```

Comment:
One of the reasons I don't like it is that it reduces readability and hinders code reuse.

If I want the answer to the question, "what functions does this file have that I could potentially reuse?", nesting functions makes that question (potentially much) more difficult to answer. Indentation makes it harder to see functions in the first place. And more importantly, nested functions can make use of variables within the enclosing function's scope. So now to figure out if I can reuse one of those nested functions, I have to look through *all* of the variables it uses to figure out if any of them belong to the parent function. Actually I generally assume that if a function is nested, it *does* make use of some local variable, because if it didn't... why would it need to be nested in the first place? You also get hit with a 4-character indentation/line-wrapping penalty, but that's a minor point :)